realm/realm-library/src/objectServer/java/io/realm/permissions/UserCondition.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.permissions;
+
+import io.realm.PermissionManager;
+import io.realm.SyncUser;
+import io.realm.internal.Util;
+
+
+/**
+ * This class represents a condition for matching users on the Realm Object Server.
+ * It is used when a request for changing existing permissions is done.
+ *
+ * @see PermissionRequest
+ * @see io.realm.PermissionManager#applyPermissions(PermissionRequest, PermissionManager.Callback)
+ */
+public final class UserCondition {
+
+    private final String key;
+    private final String value;
+    private final MatcherType type;
+
+    /**
+     * Creates a condition for matching, exactly, a users username. The comparison is case-sensitive and wildcards are
+     * not allowed.
+     *
+     * @param username exact username to match against.
+     */
+    public static UserCondition username(String username) {
+        if (Util.isEmptyString(username)) {
+            throw new IllegalArgumentException("Non-empty 'username' required.");
+        }
+        return new UserCondition(MatcherType.METADATA, "email", username);
+    }
+
+    /**
+     * Creates a condition for matching, exactly, a users id.
+     *
+     * @param userId user id to match against. No wildcards are allowed.
+     *
+     * @see SyncUser#getIdentity()
+     */
+    public static UserCondition userId(String userId) {
+        if (Util.isEmptyString(userId)) {
+            throw new IllegalArgumentException("Non-empty 'userId' required.");
+        }
+        return new UserCondition(MatcherType.USER_ID, "", userId);
+    }
+
+    /**
+     * Creates a condition that will match all users with no permissions for the Realm.
+     * <p>
+     * The {@link AccessLevel} defined alongside this condition, will also be used as the default access level
+     * for future new users that might be given access to the Realm.
+     *
+     * // FIMXE Link to makeOffer() method
+     */
+    public static UserCondition noExistingPermissions() {
+        return userId("*");
```

Comment:
Technically no, but the meaning is non-obvious, which is why this was added as a more descriptive way of supporting it. I would rather like to avoid documenting the support of `*` for `UserCondition.userId("*")`, but I can remove the line if you think it would be better?